src/main/java/org/anarres/parallelgzip/ParallelGZIPOutputStream.java

```diff
@@ -71,41 +78,47 @@ public Block(@Nonnegative int index) {
          */
         // Only on worker thread
         @Override
-        public byte[] call() throws Exception {
+        public Block call() throws IOException {
             // LOG.info("Processing " + this + " on " + Thread.currentThread());
 
             State state = STATE.get();
             // ByteArrayOutputStream buf = new ByteArrayOutputStream(in.length);   // Overestimate output size required.
             // DeflaterOutputStream def = newDeflaterOutputStream(buf);
             state.def.reset();
             state.buf.reset();
-            state.str.write(in, 0, in_length);
+            state.str.write(buf, 0, buf_length);
             state.str.flush();
 
+            buf_length = state.buf.size();
+            assert buf_length <= buf.length;
+            state.buf.writeTo(buf);
+
             // return Arrays.copyOf(in, in_length);
-            return state.buf.toByteArray();
+            return this;
         }
 
         @Override
         public String toString() {
-            return "Block" /* + index */ + "(" + in_length + "/" + in.length + " bytes)";
+            return "Block" /* + index */ + "(" + buf_length + "/" + buf.length + " bytes)";
         }
     }
     // TODO: Share, daemonize.
     private final ExecutorService executor;
     private final int nthreads;
     private final CRC32 crc = new CRC32();
-    private final BlockingQueue<Future<byte[]>> emitQueue;
-    private Block block = new Block(/* 0 */);
+    private final BlockingQueue<Future<Block>> emitQueue;
+    private ObjectArrayList<Block> blocks = new ObjectArrayList<>();  // list of reusable blocks
+    private Block block = new Block();
+    private byte[] blockBuf = block.buf;
     /** Used as a sentinel for 'closed'. */
-    private int bytesWritten = 0;
+    private long bytesWritten = 0;
 
     // Master thread only
     public ParallelGZIPOutputStream(@Nonnull OutputStream out, @Nonnull ExecutorService executor, @Nonnegative int nthreads) throws IOException {
         super(out);
         this.executor = executor;
         this.nthreads = nthreads;
-        this.emitQueue = new ArrayBlockingQueue<Future<byte[]>>(nthreads);
+        this.emitQueue = new ArrayBlockingQueue<>(nthreads);
```

Comment:
I found a note in the build file that said we could still run on 1.7, which some of our customers still require. :-(